collector/mdadm.go

```diff
@@ -0,0 +1,263 @@
+	disksActive  int64
+	disksTotal   int64
+	blocksTotal  int64
+	blocksSynced int64
+}
+
+type mdadmCollector struct {}
+
+func init() {
+	Factories["mdadm"] = NewMdadmCollector
+}
+
+func evalStatusline(statusline string) (active, total, size int64, err error) {
+
+	matches := statuslineRE.FindStringSubmatch(statusline)
+
+	// +1 to make it more obvious that the whole string containing the info is also returned as matches[0].
+	if len(matches) < 3+1 {
+		return 0, 0, 0, fmt.Errorf("too few matches found in statusline: %s", statusline)
+	} else {
+		if len(matches) > 3+1 {
+			return 0, 0, 0, fmt.Errorf("too many matches found in statusline: %s", statusline)
+		}
+	}
+
+	size, err = strconv.ParseInt(matches[1], 10, 64)
+	if err != nil {
+		return 0, 0, 0, fmt.Errorf("%s in statusline: %s", err, statusline)
+	}
+
+	total, err = strconv.ParseInt(matches[2], 10, 64)
+	if err != nil {
+		return 0, 0, 0, fmt.Errorf("%s in statusline: %s", err, statusline)
+	}
+	active, err = strconv.ParseInt(matches[3], 10, 64)
+	if err != nil {
+		return 0, 0, 0, fmt.Errorf("%s in statusline: %s", err, statusline)
+	}
+
+	return active, total, size, nil
+}
+
+// Gets the size that has already been synced out of the sync-line.
+func evalBuildline(buildline string) (int64, error) {
+	matches := buildlineRE.FindStringSubmatch(buildline)
+
+	// +1 to make it more obvious that the whole string containing the info is also returned as matches[0].
+	if len(matches) < 1+1 {
+		return 0, fmt.Errorf("too few matches found in buildline: %s", buildline)
+	} else {
+		if len(matches) > 1+1 {
+			return 0, fmt.Errorf("too many matches found in buildline: %s", buildline)
+		}
+	}
+
+	syncedSize, err := strconv.ParseInt(matches[1], 10, 64)
+
+	if err != nil {
+		return 0, fmt.Errorf("%s in buildline: %s", err, buildline)
+	} else {
+		return syncedSize, nil
+	}
+}
+
+// Parses an mdstat-file and returns a struct with the relevant infos.
+func parseMdstat(mdstatusfilepath string) ([]mdstatus, error) {
+	content, err := ioutil.ReadFile(mdstatusfilepath)
+	if err != nil {
+		return []mdstatus{}, fmt.Errorf("error parsing %s: %s", statusfile, err)
+	}
+
+	mdStatusFile := string(content)
+
+	lines := strings.Split(mdStatusFile, "\n")
+	var currentMD string
+
+	// Each md has at least the deviceline, statusline and one empty line afterwards
+	// so we will have probably something of the order len(lines)/3 devices
+	// so we use that for preallocation.
+	estimateMds := len(lines) / 3
+	mdstates := make([]mdstatus, 0, estimateMds)
+
+	for i, l := range lines {
+		if l == "" {
+			// Skip entirely empty lines.
+			continue
+		}
+
+		if l[0] == ' ' {
+			// Those lines are not the beginning of a md-section.
+			continue
+		} else {
+
+			if strings.HasPrefix(l, "Personalities") || strings.HasPrefix(l, "unused") {
+				// We aren't interested in lines with general info.
+				continue
+			}
+
+			mainLine := strings.Split(l, " ")
+			currentMD = mainLine[0]               // name of md-device
```

Comment:
On index accesses like this, please check first whether the given index exists (`if len(mainLine) < 3 {...}`). Otherwise the entire node exporter will crash on a bad line.